File: src/main/java/com/leijendary/spring/boot/template/repository/SoftDeleteRepositoryImpl.java
package com.leijendary.spring.boot.template.repository;

import static com.leijendary.spring.boot.template.util.RequestContext.USERNAME_SYSTEM;
import static com.leijendary.spring.boot.template.util.RequestContext.getUsername;
import static com.leijendary.spring.boot.template.util.RequestContext.now;
import static java.util.Optional.ofNullable;

import javax.persistence.EntityManager;

import com.leijendary.spring.boot.template.model.SoftDeleteModel;

import org.springframework.stereotype.Repository;

import lombok.RequiredArgsConstructor;

@Repository
@RequiredArgsConstructor
public class SoftDeleteRepositoryImpl<T extends SoftDeleteModel> implements SoftDeleteRepository<T> {

    private final EntityManager entityManager;

    @Override
    public void softDelete(final T entity) {
        final var username = ofNullable(getUsername())
                .orElse(USERNAME_SYSTEM);

        entity.setDeletedAt(now());
        entity.setDeletedBy(username);

        entityManager.persist(entity);
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/repository/SoftDeleteRepositoryImpl.kt
package com.leijendary.spring.template.core.repository

import com.leijendary.spring.template.core.entity.SoftDeleteEntity
import com.leijendary.spring.template.core.util.RequestContext.now
import com.leijendary.spring.template.core.util.RequestContext.userId
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import org.springframework.stereotype.Repository

@Repository
class SoftDeleteRepositoryImpl<T : SoftDeleteEntity>(
    @PersistenceContext
    private val entityManager: EntityManager
) : SoftDeleteRepository<T> {

    override fun softDelete(entity: T) {
        entity.deletedAt = now
        entity.deletedBy = userId

        entityManager.persist(entity)
    }
}
